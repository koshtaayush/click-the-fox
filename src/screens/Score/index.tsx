//Library
import React from 'react'
import styled from 'styled-components'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'

//Shared interface
import { playerPerformance } from './../../typings/sharedInterfaces'

//Utils
import { formatDate } from './../../utils/formatDate'

//Reusable Components
import GameHeading from './../../components/GameHeading'
import Button from '../../components/Button'

//Actions
import {
    fetchImages,
    setPlayerPlaying,
    toggleIsPlayClicked,
} from './../../actions/appActions/app.actions'

interface Props {}

const Score: React.FC<Props> = () => {
    const { overallStanding } = useSelector(
        (state: {
            app: {
                overallStanding: Array<playerPerformance>
            }
        }) => ({
            overallStanding: state.app.overallStanding,
        })
    )

    const dispatch = useDispatch()
    const history = useHistory()

    /**
     * Function to move user to welcome screen
     */
    const handleMoveToWelcomeScreen = () => {
        dispatch(toggleIsPlayClicked())
        dispatch(setPlayerPlaying(''))
        history.push('/')
    }

    /**
     * Function to show new set of images to user and play again
     */
    const handlePlayAgain = () => {
        dispatch(fetchImages())
        history.push('/play')
    }

    return (
        <React.Fragment>
            <Container>
                <GameHeading />
                <Label>Scoreboard</Label>
                <ScoreTable>
                    <thead>
                        <TableRow>
                            <TableLegend>Rank</TableLegend>
                            <TableLegend>Name</TableLegend>
                            <TableLegend>Date</TableLegend>
                            <TableLegend>Score</TableLegend>
                        </TableRow>
                    </thead>
                    <tbody>
                        {overallStanding.map(
                            (playerScore: playerPerformance, index: number) => {
                                return (
                                    <TableRow key={index} test-id="scoreTableRow">
                                        <RowRank>{index + 1}</RowRank>
                                        <RowInfo
                                            index={index}
                                            textAlignment={'start'}>
                                            {playerScore.name}
                                        </RowInfo>
                                        <RowInfo
                                            index={index}
                                            textAlignment={'center'}>
                                            {formatDate(playerScore.datePlayed)}
                                        </RowInfo>
                                        <RowInfo
                                            index={index}
                                            textAlignment={'end'}>
                                            {playerScore.score}
                                        </RowInfo>
                                    </TableRow>
                                )
                            }
                        )}
                    </tbody>
                </ScoreTable>
                <BottomActions>
                    <MoveToWelcomeScreen>
                        <Button
                            onClickProp={handleMoveToWelcomeScreen}
                            buttonText={'Move to Welcome screen'}
                            disabled={false}
                            buttonType={'S'}
                        />
                    </MoveToWelcomeScreen>

                    <PlayAgain>
                        <Button
                            onClickProp={handlePlayAgain}
                            buttonText={'PLAY!'}
                            disabled={false}
                            buttonType={'L'}
                        />
                    </PlayAgain>
                </BottomActions>
            </Container>
        </React.Fragment>
    )
}

export default Score

interface IRowInfo {
    index: number
    textAlignment: string
}

const Container = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    flex-direction: column;
    height: calc(100% - 4rem);
    width: calc(100% - 4rem);
`

const BottomActions = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    justify-content: space-evenly;
    width: 50%;
`
const MoveToWelcomeScreen = styled.div``
const PlayAgain = styled.div``
const Label = styled.div`
    font-weight: 500;
    font-size: 1.5rem;
    text-transform: uppercase;
`

const ScoreTable = styled.table`
    font-weight: 500;
    min-width: 40%;
`

const TableLegend = styled.th`
    background-color: #a5a5a5;
    color: #ffffff;
    padding: 1rem;
`

const RowRank = styled.td`
    background-color: #a5a5a5;
    color: #ffffff;
    padding: 1rem;
    text-align: center;
`

const RowInfo = styled.td<IRowInfo>`
    background-color: ${(props) =>
        props.index % 2 === 0 ? '#E1E1E1' : '#F0F0F0'};
    color: #ffffff;
    padding: 1rem;
    color: #000;
    font-weight: 500;
    text-align: ${(props) => props.textAlignment};
`

const TableRow = styled.tr``
