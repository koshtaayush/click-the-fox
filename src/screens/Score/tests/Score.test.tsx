import React from 'react'
import { shallow } from 'enzyme'
import * as Redux from 'react-redux'

import Score from './../index'

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: jest.fn(),
    }),
}))

describe('Score', () => {

    const defaultState = {
        app: {
            overallStanding: []
        },
    }

    const setUpUseSelector = (state = {}) => {
        const localState = {
            app: { ...defaultState.app, ...state },
        }
        const setupState = { ...localState }
        const useSelectorMock = jest.spyOn(Redux, 'useSelector')
        return useSelectorMock.mockImplementation((callback) =>
            callback(setupState)
        )
    }

    const setUpUseDispatch = (state = {}) => {
        const useDispatchMock = jest.spyOn(Redux, 'useDispatch')
        const mockDispatch = jest.fn()
        return useDispatchMock.mockReturnValue(mockDispatch)
    }

    beforeEach(() => {
        setUpUseDispatch()
    })

    it('Should match the snapshot', () => {

        setUpUseSelector()

        const component = shallow(
            <Score />
        )

        expect(component).toMatchSnapshot()
    })
    
    it('Should render the correct number of table rows', () => {

        setUpUseSelector({
            overallStanding: [
                {
                  name: 'Player 1',
                  datePlayed: new Date(),
                  score: 2
                },
                {
                  name: 'Player 1',
                  datePlayed: new Date(),
                  score: 1
                }
              ]
        })

        const component = shallow(
            <Score />
        )

        expect(component.find({'test-id': 'scoreTableRow'})).toHaveLength(2)
    })
})
