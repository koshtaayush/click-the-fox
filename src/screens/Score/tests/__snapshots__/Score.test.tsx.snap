// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Score Should match the snapshot 1`] = `
<Fragment>
  <styled.div>
    <GameHeading />
    <styled.div>
      Scoreboard
    </styled.div>
    <styled.table>
      <thead>
        <styled.tr>
          <styled.th>
            Rank
          </styled.th>
          <styled.th>
            Name
          </styled.th>
          <styled.th>
            Date
          </styled.th>
          <styled.th>
            Score
          </styled.th>
        </styled.tr>
      </thead>
      <tbody />
    </styled.table>
    <styled.div>
      <styled.div>
        <Button
          buttonText="Move to Welcome screen"
          buttonType="S"
          disabled={false}
          onClickProp={[Function]}
        />
      </styled.div>
      <styled.div>
        <Button
          buttonText="PLAY!"
          buttonType="L"
          disabled={false}
          onClickProp={[Function]}
        />
      </styled.div>
    </styled.div>
  </styled.div>
</Fragment>
`;
